mergify_engine/merge_train.py

```diff
@@ -0,0 +1,440 @@
+import redis
+
+from mergify_engine import check_api
+from mergify_engine import constants
+from mergify_engine import context
+from mergify_engine import github_events
+from mergify_engine import github_types
+from mergify_engine import rules
+from mergify_engine import subscription
+from mergify_engine import utils
+from mergify_engine.clients import github
+from mergify_engine.clients import http
+
+
+class WaitingPull(typing.NamedTuple):
+    pull_request_number: github_types.GitHubPullRequestNumber
+    queue_name: rules.QueueName
+
+
+@dataclasses.dataclass
+class TrainCar:
+    train: "Train" = dataclasses.field(repr=False)
+    pull_request_number: github_types.GitHubPullRequestNumber
+    queue_name: rules.QueueName
+    parents: typing.List[github_types.GitHubPullRequestNumber]
+    initial_current_base_sha: github_types.SHAType
+    current_base_sha: github_types.SHAType
+    tmp_pull_number: typing.Optional[
+        github_types.GitHubPullRequestNumber
+    ] = dataclasses.field(default=None)
+
+    class Serialized(typing.TypedDict):
+        pull_request_number: github_types.GitHubPullRequestNumber
+        queue_name: rules.QueueName
+        parents: typing.List[github_types.GitHubPullRequestNumber]
+        initial_current_base_sha: github_types.SHAType
+        current_base_sha: github_types.SHAType
+        tmp_pull_number: typing.Optional[github_types.GitHubPullRequestNumber]
+
+    def serialized(self) -> "TrainCar.Serialized":
+        return self.Serialized(
+            pull_request_number=self.pull_request_number,
+            queue_name=self.queue_name,
+            parents=self.parents,
+            initial_current_base_sha=self.initial_current_base_sha,
+            current_base_sha=self.current_base_sha,
+            tmp_pull_number=self.tmp_pull_number,
+        )
+
+    @classmethod
+    def deserialized(cls, train: "Train", data: "TrainCar.Serialized") -> "TrainCar":
+        return cls(train, **data)
+
+    def create_pull(self) -> None:
+        # TODO(sileht): reuse branch instead of recreating PRs ?
+        branch_name = f"{constants.MERGE_QUEUE_BRANCH_PREFIX}/{self.train.ref}/{self.pull_request_number}"
+
+        self.train.client.post(
+            f"/repos/{self.train.owner}/{self.train.repo}/git/refs",
+            json={
+                "ref": f"refs/heads/{branch_name}",
+                "sha": self.initial_current_base_sha,
+            },
+        )
+
+        for pull_number in self.parents + [self.pull_request_number]:
+            # TODO(sileht): if a merge fail we should update the summary of self.pull_request_number with
+            # the failure
+            self.train.client.post(
+                f"/repos/{self.train.owner}/{self.train.repo}/merges",
+                json={
+                    "base": branch_name,
+                    "head": f"refs/pull/{pull_number}/head",
+                    "commit_message": f"Merge of #{pull_number}",
+                },
+            )
```

Comment:
```suggestion
        self.train.client.post(
            f"/repos/{self.train.owner}/{self.train.repo}/git/refs",
            json={
                "ref": f"refs/heads/{branch_name}",
                "sha": self.initial_current_base_sha,
            },
        )

        for pull_number in self.parents + [self.pull_request_number]:
            # TODO(sileht): if a merge fail we should update the summary of self.pull_request_number with
            # the failure
            self.train.client.post(
                f"/repos/{self.train.owner}/{self.train.repo}/merges",
                json={
                    "base": branch_name,
                    "head": f"refs/pull/{pull_number}/head",
                    "commit_message": f"Merge of #{pull_number}",
                },
            )
```
Do we need to handle failure?

Reply:
That's the last remaining TODO and test cases to write :)

Reply:
For asyncio, I want to update check_api first.